app_test.go

```diff
@@ -1428,6 +1428,275 @@ func TestValidateApp(t *testing.T) {
 	})
 }
 
+func TestHookConstructors(t *testing.T) {
+	t.Run("all", func(t *testing.T) {
+		var (
+			calls = map[string]int{
+				"start func":         0,
+				"start func err":     0,
+				"start ctx func":     0,
+				"start ctx func err": 0,
+				"stop func":          0,
+				"stop func err":      0,
+				"stop ctx func":      0,
+				"stop ctx func err":  0,
+			}
+			nilFunc             func()
+			nilErrorFunc        func() error
+			nilContextFunc      func(context.Context)
+			nilContextErrorFunc func(context.Context) error
+		)
+
+		fxtest.New(
+			t,
+			Invoke(func(lc Lifecycle) {
+				// Nil functions
+				lc.Append(StartStopHook(nilFunc, nilErrorFunc))
+				lc.Append(StartStopHook(nilContextFunc, nilContextErrorFunc))
+
+				// Start hooks
+				lc.Append(StartHook(func() {
+					calls["start func"]++
+				}))
+				lc.Append(StartHook(func() error {
+					calls["start func err"]++
+					return nil
+				}))
+				lc.Append(StartHook(func(context.Context) {
+					calls["start ctx func"]++
+				}))
+				lc.Append(StartHook(func(context.Context) error {
+					calls["start ctx func err"]++
+					return nil
+				}))
+
+				// Stop hooks
+				lc.Append(StopHook(func() {
+					calls["stop func"]++
+				}))
+				lc.Append(StopHook(func() error {
+					calls["stop func err"]++
+					return nil
+				}))
+				lc.Append(StopHook(func(context.Context) {
+					calls["stop ctx func"]++
+				}))
+				lc.Append(StopHook(func(context.Context) error {
+					calls["stop ctx func err"]++
+					return nil
+				}))
+
+				// StartStop hooks
+				lc.Append(StartStopHook(
+					func() {
+						calls["start func"]++
+					},
+					func() error {
+						calls["stop func err"]++
+						return nil
+					},
+				))
+				lc.Append(StartStopHook(
+					func() error {
+						calls["start func err"]++
+						return nil
+					},
+					func(context.Context) {
+						calls["stop ctx func"]++
+					},
+				))
+				lc.Append(StartStopHook(
+					func(context.Context) {
+						calls["start ctx func"]++
+					},
+					func(context.Context) error {
+						calls["stop ctx func err"]++
+						return nil
+					},
+				))
+				lc.Append(StartStopHook(
+					func(context.Context) error {
+						calls["start ctx func err"]++
+						return nil
+					},
+					func() {
+						calls["stop func"]++
+					},
+				))
+			}),
+		).RequireStart().RequireStop()
+
+		for name, count := range calls {
+			require.Equalf(t, 2, count, "bad call count: %s (%d)", name, count)
+		}
+	})
+
+	t.Run("start errors", func(t *testing.T) {
```

Comment:
I think this, along with "stop errors", can be combined into a [table test](https://github.com/uber-go/guide/blob/master/style.md#test-tables).

Reply:
Big fan of table tests, however none of the other tests here are table tests, so I'd prefer to keep this consistent with other tests as part of this change.

Fx tests are also not a great fit for table tests as currently written. We could come back and refactor/restructure them later, if we decide that's worthwhile.